File: typescript/pascals-triangle/pascals-triangle.ts
export default class Triangle {
    rows: number[][] = []
    lastRow: number[] = []

    constructor(n: number) {
        function value(position: number, row: number): number {
            if (position == 0 || position == row - 1) return 1
            else if (position == row) return 0
            else {
                return value(position - 1, row - 1) + value(position, row - 1)
            }
        }

        for (let i = 1; i <= n; i++) {
            let row = []
            for (let j = 0; j < i; j++) {
                row.push(value(j, i))
            }

            this.rows.push(row)
        }

        this.lastRow = this.rows[n - 1]
    }
}

File: scala/pascals-triangle/src/main/scala/PascalsTriangle.scala
object PascalsTriangle {
  def rows(n: Int): List[List[Int]] = {
    def triangle(current: Int, rows: List[List[Int]]): List[List[Int]] = {
      if (current >= n) rows.reverse
      else {
        val row = (1 :: rows.headOption.getOrElse(Nil).sliding(2).map(_.sum).toList) :+ 1
        triangle(current + 1, row :: rows)
      }
    }

    n match {
      case _ if n <= 0 => Nil
      case _ => triangle(2, List(List(1, 1), List(1))).take(n)
    }
  }
}
